keps/sig-node/3721-support-for-env-files/README.md

```diff
@@ -0,0 +1,560 @@
+and `FileKeySelector` field under [`EnvVarSource`](https://pkg.go.dev/k8s.io/api/core/v1#EnvVarSource)
+that can be set for an individual container.
+
+### Pod API
+
+`FileEnvSource` field will allow the users to populate environment variables
+from the specified file. The user can specify a file using `VolumeName` and
+`Path` fields. The `VolumeName` field specifies the volume mount that contains
+the file and `Path` is the relative path in this volume mount filesystem.
+
+```
+type EnvFromSource struct {
+    ...
+    // The file to select from
+    // +optional
+    FileRef *FileEnvSource `json:"fileRef,omitempty" protobuf:"bytes,3,opt,name=fileRef"`
+    ...
+}
+
+type FileEnvSource struct {
+    // The name of the volume mount containing the env file.
+    VolumeName string `json:",inline" protobuf:"bytes,1,opt,name=volumeName"`
+    // The relative file path inside the volume mount to select from.
+    Path string `json:",inline" protobuf:"bytes,2,opt,name=path"`
+    // Specify whether the file must exist.
+    // +optional
+    Optional *bool `json:"optional,omitempty" protobuf:"varint,3,opt,name=optional"`
+}
+```
+
+`FileKeySelector` field will allow the users to select the value for environment
+variable from the specified key in the file.
+
+```
+type EnvVarSource struct {
+    ...
+    // Selects a key of the env file.
+    // +optional
+    FileKeyRef *FileKeySelector `json:"fileKeyRef,omitempty" protobuf:"bytes,5,opt,name=fileKeyRef"`
+    ...
+}
+
+type FileKeySelector struct {
+    // The name of the volume mount containing the env file.
+    VolumeName string `json:",inline" protobuf:"bytes,1,opt,name=volumeName"`
+    // The relative file path inside the volume mount to select from.
+    Path string `json:",inline" protobuf:"bytes,2,opt,name=path"`
+    // The key of the env file to select from.  Must be a valid key.
+    Key string `json:"key" protobuf:"bytes,3,opt,name=key"`
+    // Specify whether the file or its key must be defined
+    // +optional
+    Optional *bool `json:"optional,omitempty" protobuf:"varint,4,opt,name=optional"`
+}
+```
+
+### Env File
+
+The full specification of an env file:
+
+1. **File Format**: The environment variable (env) file must adhere to valid YAML syntax to ensure correct parsing.
+
+2. **Variable Naming**:
+    
+    a. **Standard**: Environment variable names should be composed of alphanumeric characters (letters and numbers), underscores, dots, or hyphens. The first character cannot be a digit.
+
+    b. **Relaxed**: If the Kubernetes feature gate `RelaxedEnvironmentVariableValidation` is enabled, the naming restriction is less stringent, allowing any printable ASCII character except the equals sign (=).
+
+3. **Duplicate Names**: If an environment variable is defined multiple times in the file, the last occurrence takes precedence and overrides any previous values.
```

Comment:
We need to specify precedence wrt the pod itself.  I am worried that this becomes a new attack vector - if I `kubectl get pod` I would see `FOO=clean`, but sneaky attacker overwrote my env file and set `FOO=dirty`.  That could meaningfully compromise my application.  Yes, I opened the door to it, but Kube enabled it to happen.

I am loathe to send back all of the env vars in status, but that would be the only way to even _know_, and that is a poor monitoring mechanism.  And yes, reading multiple keys from secrets/configmaps has this same weakness.  

Typing this out, I really don't think we should do it, at least not as described.  FileKeyRef is OK.  What if we stopped there?  Or what if we required a key-prefix on FileRef (so we only load env that starts with ...)?

I am open to ideas here, up to and including "don't do it".  Apps CAN work around this themselves by wrapping in a shell script.  Did we detail why that is not a viable alternative?

Reply:
> I am open to ideas here, up to and including "don't do it". Apps CAN work around this themselves by wrapping in a shell script. Did we detail why that is not a viable alternative?

The alternative is detailed here: https://github.com/kubernetes/kubernetes/issues/116993 And this KEP meant to be a good and safe way to avoid wrapping container execution with `source` command - quality of life improvement. This is one of the reasons I was against the use case that was added initially by getting env vars from host path without even mounting it.

I see an issue with the race condition - definitely something that should be looked at. But security issue is nothing new and situation will be even worse with implementations that wrap executables and/or using NRI plugins to inject environment variables.

Reply:
From the roadmap perspective I am wrapping my head around how much a Pod should become a piepeline or orchestrator of containers. This KEP creates the ordering requirement. Pod-level requests/limits KEP creates an immediate question on how to evict "bad" container instead of OOM-ing the whole Pod. There should be some boundaries long term in what we allow to orchestrate and what we don't. This KEP is less scary for me from that perspective as it doesn't introduce a new dimension of "orchestration". We already have numerous Pods that depend on containers startup ordering. So this is just a quality of life simplification.

Reply:
We can add a pod validation check - if a container is consuming env file from an `emptyDir` volume, then we only allow mounting this `emptyDir` volume (containing the env file) in initContainers.

Reply:
This is only loading the variables specifically named in the pod spec, right? Not the whole env file? In that case, we can just have validation that prevents environment variables from being set by multiple sources.

Reply:
I still like to limit mounting of this file to init containers (not sidecars). This will eliminate any race conditions

Reply:
How do we know that a given volume is intended to hold an env file?  I don't like adding lots of restrictions on "normal" things, just because they are being used in a new way.  A volume is a volume.  We shouldn't add new restrictions on mounting it, just clear documentation about it.  Adding the "only emptyDir" restriction on the CONSUMER side makes sense.

Reply:
The `fileKeyRef` spec tells us which volume contains the env file. We can use this to check that the volume is emptyDir and only mounted in initContainer.

```
env:
    - name: CONFIG_VAR
      valueFrom:
        fileKeyRef:
          path: config.env
          volumeName: config
          key: CONFIG_VAR
```

Reply:
For us it should be nice to load environment variables from file. Currently we are using an entrypoint to retrieve some secrets and configurations, create the env file so the main container could do a "source" of the file in some entrypoint.sh script. Now, we want to start using "distroless" images, so no shell is included, so we cannot source the file. That's why it should be great for us to include this solution of loading envs from file.

Reply:
distroless is a great motivation. Good example.

I am not sure if @AnishShah plans to work on it going forward. I think this may be a good quality of life improvement for sure

Reply:
Thanks. I added this to the possible use-case above.